R/data.R

```diff
@@ -168,6 +167,56 @@ track_data <- function(
   )
 }
 
+
+#' Data object builder for a csv file
+#'
+#' Build the data object for gosling plots
+#'
+#' @param file A character. Specify the URL address or local file name in the www directory of the data file.
+#' @param genomicFields A character vector. Specify the name of genomic data fields.
+#' @param chromosomeField A character. Specify the name of chromosome data fields.
+#' @param separator A character. Specify file separator, Default: ','
+#' @param sampleLength A number. Specify the number of rows loaded from
+#' the URL. Default: 1000
+#' @param headerNames A character vector. Specify the names of data fields
+#' if a CSV file does not have header row.
+#' @param ... Any other parameters passed to json data object.
+#'
+#' @return list of data specs for a csv file
+#' @export
+track_data_csv <- function(
+    file, genomicFields = NULL, chromosomeField = NULL,
+    separator = ",", sampleLength = 1000, headerNames = NULL, ...) {
+  list_rm_null(
+    list(
+      url = file, type = "csv",
+      genomicFields = genomicFields, chromosomeField = chromosomeField,
+      separator = separator, sampleLength = sampleLength, headerNames = headerNames, ...
+    )
+  )
+}
+
+#' Data object builder for a GRanges object by locally saving it
+#'
+#' Build the data object for gosling plots
+#'
+#' @param granges A GRanges object from the GenomicRanges package with seqnames and ranges
+#'
+#' @return list of data specs for a csv file
+#' @export
+track_data_gr <- function(granges) {
+  file_name <- paste0(digest::digest(granges), ".csv")
+  if (!file_name %in% list.files(".gosling")) {
+    utils::write.csv(granges, paste0(".gosling/", file_name), row.names = FALSE)
```

Comment:
This means that we will write this csv file to the `.gosling` folder within the working directory of the user? 🤔 
I can see some potential problems and side effects that might be annoying for the end users by doing this.
Is there a reason for us not writing to `tempdir()`?
What if we have a couple of helper functions like:
```r
write_temp_csv <- function(filename, data) {
  utils::write.csv(data, filename, row.names = FALSE)
}

csv_temp_file_exists  <- function(filepath) {
  file.exists(filepath)
}

get_or_create_granges_csv <- function(granges) {
  filename <- digest::digest(granges)
  filepath <- tempfile(pattern = filename, filet = ".csv")
  if (!csv_temp_file_exists(filepath)) {
    write_temp_csv(filepath)
  }
  filepath
}

Reply:
And later we can
```R
  granges_file_name <- get_or_create_granges_csv(granges)
  track_data_csv(
     file = granges_file_name,
     chromosomeField = "seqnames",
     genomicFields = c("start", "end")
   )
```

Reply:
Thanks @federiva it's a good suggestion.
However, I was unable to make it work this way as the csv file needs to be accessible by the shiny app's internal static server. Additionally, The reason for choosing a local directory is that sometimes the csv file is too big and it can take a while to save it, and having it in the project directory will make sure that we can reuse the saved file (another reason why we use hashed file names if the content is not changed). Using a temp directory would create a temp location that is not accessible by a new shiny session.
<img width="684" alt="Screenshot 2023-06-27 at 11 24 03 AM" src="https://github.com/Appsilon/shiny.gosling/assets/49812166/93944ff9-2723-4eaa-abb5-67577969940e">

Reply:
The key thing why the `.gosling` directory works is because we're making it a directory for static files:
```
shiny::addResourcePath(".gosling", fs::path_wd(".gosling"))
```
So we're able to access the csv files inside it by their hash names.
<img width="681" alt="Screenshot 2023-06-27 at 11 30 21 AM" src="https://github.com/Appsilon/shiny.gosling/assets/49812166/890c4c68-872f-44a8-b39b-e11c773a61f6">

Reply:
Btw if the file is very big, will the browser become heavy and laggy ? Did we test this ?
We need to check this with Kuba N. There can be many issues with such a setup. Although I suggested it before but thinking about it now, I notice the issues.
@vedhav Can you please check with @jakubnowicki . Maybe there is a btter simpler solution to this.